pkg/geo/geomfn/generate_points.go

```diff
@@ -0,0 +1,214 @@
+
+package geomfn
+
+import (
+	"math"
+	"math/rand"
+
+	"github.com/cockroachdb/cockroach/pkg/geo"
+	"github.com/cockroachdb/cockroach/pkg/geo/geopb"
+	"github.com/cockroachdb/cockroach/pkg/geo/geos"
+	"github.com/cockroachdb/errors"
+	"github.com/twpayne/go-geom"
+)
+
+// GenerateRandomPoints generates provided number of pseudo-random points for the input area.
+func GenerateRandomPoints(g geo.Geometry, nPoints int, rng *rand.Rand) (geo.Geometry, error) {
+	if nPoints < 0 {
+		return geo.Geometry{}, nil
+	}
+	pointsAsGeometry, err := generateRandomPoints(g, nPoints, rng)
+	if err != nil {
+		return geo.Geometry{}, errors.Wrap(err, "generating random points error")
+	}
+	return pointsAsGeometry, nil
+}
+
+// generateRandomPoints returns a MultiPoint geometry consisting of randomly generated points
+// that are covered by the geometry provided.
+// nPoints is the number of points to return.
+// rng is the random numbers generator.
+func generateRandomPoints(g geo.Geometry, nPoints int, rng *rand.Rand) (geo.Geometry, error) {
+	var generateRandomPointsFunction func(g geo.Geometry, nPoints int, rng *rand.Rand) (*geom.MultiPoint, error)
+	switch g.ShapeType() {
+	case geopb.ShapeType_Polygon:
+		generateRandomPointsFunction = generateRandomPointsFromPolygon
+	case geopb.ShapeType_MultiPolygon:
+		generateRandomPointsFunction = generateRandomPointsFromMultiPolygon
+	default:
+		return geo.Geometry{}, errors.Newf("unsupported type: %v", g.ShapeType().String())
+	}
+	// This is to be checked once we know Geometry type is supported,
+	// so that we can keep consistency with PostGIS implementation.
+	if nPoints == 0 {
+		return geo.Geometry{}, nil
+	}
+	empty, err := IsEmpty(g)
+	if err != nil {
+		return geo.Geometry{}, errors.Wrap(err, "could not check if geometry is empty")
+	}
+	if empty {
+		return geo.Geometry{}, nil
+	}
+	mpt, err := generateRandomPointsFunction(g, nPoints, rng)
+	if err != nil {
+		return geo.Geometry{}, err
+	}
+	srid := g.SRID()
+	mpt.SetSRID(int(srid))
+	out, err := geo.MakeGeometryFromGeomT(mpt)
+	if err != nil {
+		return geo.Geometry{}, errors.Wrap(err, "could not transform geom.T into geometry")
+	}
+	return out, nil
+}
+
+func generateRandomPointsFromPolygon(
+	g geo.Geometry, nPoints int, rng *rand.Rand,
+) (*geom.MultiPoint, error) {
+	area, err := Area(g)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not calculate Polygon area")
+	}
+	if area == 0.0 {
+		return nil, errors.New("zero area input Polygon")
+	}
+	bbox := g.CartesianBoundingBox()
+	bboxWidth := bbox.HiX - bbox.LoX
+	bboxHeight := bbox.HiY - bbox.LoY
+	bboxArea := bboxHeight * bboxWidth
+	// Gross up our test set a bit to increase odds of getting coverage in one pass.
+	sampleNPoints := float64(nPoints) * bboxArea / area
+
+	sampleSqrt := math.Round(math.Sqrt(sampleNPoints))
+	if sampleSqrt == 0 {
+		sampleSqrt = 1
+	}
+	var sampleHeight, sampleWidth int
+	var sampleCellSize float64
+	// Calculate the grids we're going to randomize within.
```

Comment:
This needs some commenting for the next person to look at this. What is a grid in this instance, and what are sample heights / widths? Maybe a hollistic view of the algorithm

Reply:
Ok, added function description.